include/micm/solver/cuda_lu_decomposition.hpp

```diff
@@ -11,26 +11,85 @@
   class CudaLuDecomposition : public LuDecomposition
   {
-   public:
-    CudaLuDecomposition(){};
+    public:
+      /// This is an instance of struct "LuDecomposeConst" that holds
+      ///   the constant data of "CudaLuDecomposition" class on the device
+      LuDecomposeConst* devptr;
 
-    template<typename T, typename OrderingPolicy>
-    CudaLuDecomposition(const SparseMatrix<T, OrderingPolicy>& matrix)
-        : LuDecomposition(matrix){};
+      /// This is the default constructor, taking no arguments;
+      CudaLuDecomposition(){};
 
-    template<typename T, template<class> typename SparseMatrixPolicy>
-    requires VectorizableSparse<SparseMatrixPolicy<T>> std::chrono::nanoseconds
-    Decompose(const SparseMatrixPolicy<T>& A, SparseMatrixPolicy<T>& L, SparseMatrixPolicy<T>& U)
-    const;
+      /// This is the overloaded constructor that takes one argument called "matrix";
+      /// We need to specify the type (e.g., double, int, etc) and 
+      ///   ordering (e.g., vector-stored, non-vector-stored, etc) of the "matrix";
+      template<typename T, typename OrderingPolicy>
+      CudaLuDecomposition(const SparseMatrix<T, OrderingPolicy>& matrix)
+          : LuDecomposition(matrix)
+      {
+        /// Passing the class itself as an argument is not support by CUDA;
+        /// Thus we generate a host struct first to save the pointers to 
+        ///   the actual data and size of each constant data member;
+
+        /// Allocate host memory space for an object of type "LuDecomposeConst"
+        LuDecomposeConst* hostptr = new LuDecomposeConst;
+        hostptr->niLU_         = this->niLU_.data();
+        hostptr->do_aik_       = this->do_aik_.data();
+        hostptr->aik_          = this->aik_.data();
+        hostptr->uik_nkj_      = this->uik_nkj_.data();
+        hostptr->lij_ujk_      = this->lij_ujk_.data();
+        hostptr->do_aki_       = this->do_aki_.data();
+        hostptr->aki_          = this->aki_.data();
+        hostptr->lki_nkj_      = this->lki_nkj_.data();
+        hostptr->lkj_uji_      = this->lkj_uji_.data();
+        hostptr->uii_          = this->uii_.data();
+        hostptr->niLU_size_    = this->niLU_.size();
+        hostptr->do_aik_size_  = this->do_aik_.size();
+        hostptr->aik_size_     = this->aik_.size();
+        hostptr->uik_nkj_size_ = this->uik_nkj_.size();
+        hostptr->lij_ujk_size_ = this->lij_ujk_.size();
+        hostptr->do_aki_size_  = this->do_aki_.size();
+        hostptr->aki_size_     = this->aki_.size();
+        hostptr->lki_nkj_size_ = this->lki_nkj_.size();
+        hostptr->lkj_uji_size_ = this->lkj_uji_.size();
+        hostptr->uii_size_     = this->uii_.size(); 
+        
+        // Copy the data from host struct to device struct
+        micm::cuda::CopyConstData(hostptr, this->devptr);
+
+        /// Free the host memory for "hostptr"
+        delete hostptr;
```

Comment:
```suggestion
        LuDecomposeConst hostptr = LuDecomposeConst;
        hostptr->niLU_         = this->niLU_.data();
        hostptr->do_aik_       = this->do_aik_.data();
        hostptr->aik_          = this->aik_.data();
        hostptr->uik_nkj_      = this->uik_nkj_.data();
        hostptr->lij_ujk_      = this->lij_ujk_.data();
        hostptr->do_aki_       = this->do_aki_.data();
        hostptr->aki_          = this->aki_.data();
        hostptr->lki_nkj_      = this->lki_nkj_.data();
        hostptr->lkj_uji_      = this->lkj_uji_.data();
        hostptr->uii_          = this->uii_.data();
        hostptr->niLU_size_    = this->niLU_.size();
        hostptr->do_aik_size_  = this->do_aik_.size();
        hostptr->aik_size_     = this->aik_.size();
        hostptr->uik_nkj_size_ = this->uik_nkj_.size();
        hostptr->lij_ujk_size_ = this->lij_ujk_.size();
        hostptr->do_aki_size_  = this->do_aki_.size();
        hostptr->aki_size_     = this->aki_.size();
        hostptr->lki_nkj_size_ = this->lki_nkj_.size();
        hostptr->lkj_uji_size_ = this->lkj_uji_.size();
        hostptr->uii_size_     = this->uii_.size(); 
        
        // Copy the data from host struct to device struct
        micm::cuda::CopyConstData(&hostptr, this->devptr);
```

Since it's a temporary object, there's no reason to manage the memory ourselves; a stack variable seems fine to me here

Reply:
So if we make it a stack variable, then shall we access it through something like `hostptr.niLU_`?

Reply:
yes

Reply:
Done.